features/lwipstack/lwip/src/apps/altcp_tls/altcp_tls_mbedtls.c

```diff
@@ -599,8 +599,15 @@ altcp_mbedtls_setup(void *conf, struct altcp_pcb *conn, struct altcp_pcb *inner_
     altcp_mbedtls_free(conf, state);
     return ERR_MEM;
   }
+  // Defines MBEDTLS_SSL_CONF_RECV/SEND/RECV_TIMEOUT define global functions which should be the same for all
+  // callers of mbedtls_ssl_set_bio_ctx and there should be only one ssl context. If these rules don't apply,
```

Comment:
If there is only one SSL context, have we verified that the proper synchronization is in place, if any is required, to ensure accesses to the context are threadsafe and that we avoid multithreading bugs due to multiple concurrent uses of the SSL context?

Reply:
@jarvte Can you comment?

Reply:
My comments were primarily about the single SSL context, rather than the `send()` and `recv()` functions.

Previously, one one only need to worry about locking around contexts that are shared between threads. Users often made their own contexts for use within a single thread and didn't have to worry. Now that we have one global SSL context, we need to understand if our use of the single SSL context is thread safe?

> If the app developer defines global functions to be used (outside of MbedTLS) s/he should take care of the handling of those as well.

But we are making these global functions in this case, not the app developer, right? So shouldn't we take care? We are providing a global default, not configurable, `send()` and `recv()` implementation to reduce code size. Are our `send()` and `recv()` thread safe? Do they need to be?

Reply:
In config.h (similarly for others): 
* MBEDTLS_SSL_CONF_SEND must evaluate to the name of an **externally**
 *   defined function with signature
 *      int f_send( void*, const unsigned char *, size_t )
There are definitions for those functions, but implementation needs to be done by app developer.
If I would design a system with 1 entry point for SSL connectivity(socket) functionality, then SSL context would be handled only in that same place. Then multithreading issues would be also handled in the same place with proper locks (if there is a need for more than 1 SSL context).

Maybe we can handle these cases by documenting those properly. What do you think @Patater and @sbutcher-arm?

Reply:
After discussion with @JanneKiiskila, we've agreed to limit Mbed TLS baremetal branch use with the Mbed OS baremetal build mode. This will allow us to prioritize and address multithreading issues, like this one, separately.